File: Mobile App/lib/data/services/qr_service.dart
import 'dart:io';
import 'package:supabase_flutter/supabase_flutter.dart';
import 'package:noble_invoice/core/services/supabase_service.dart';
import 'package:noble_invoice/data/models/qr_code_model.dart';
import 'package:noble_invoice/data/models/folder_model.dart';

class QrService {
  static String _parseError(dynamic e) {
    if (e is PostgrestException) return e.message;
    if (e is AuthException) return e.message;
    if (e is StorageException) return e.message;
    return e.toString();
  }

  /// Fetch all QR codes for the current user.
  static Future<List<QrCode>> getQrCodes({String? teamId}) async {
    try {
      final userId = SupabaseService.currentUser?.id;
      if (userId == null) throw Exception('Unauthorized');
      
      var query = SupabaseService.client
          .from('qr_codes')
          .select('*')
          .eq('user_id', userId);
          
      if (teamId != null) {
        query = query.eq('team_id', teamId);
      }

      final data = await query.order('created_at', ascending: false);
          
      return data.map((m) => QrCode.fromMap(m)).toList();
    } catch (e) {
      throw _parseError(e);
    }
  }

  /// Save a new QR code metadata to the database.
  static Future<String> saveQrCode(QrCode qrCode) async {
    try {
      final userId = SupabaseService.currentUser?.id;
      final map = qrCode.toMap();
      map['user_id'] = userId;
      
      final res = await SupabaseService.client
          .from('qr_codes')
          .upsert(map)
          .select('id')
          .single();
          
      return res['id']?.toString() ?? '';
    } catch (e) {
      throw _parseError(e);
    }
  }

  /// Upload an asset (PDF, MP3, Image) for a QR code.
  static Future<Map<String, dynamic>> uploadAsset(File file) async {
    try {
      final userId = SupabaseService.currentUser?.id;
      final fileName = '${DateTime.now().millisecondsSinceEpoch}_${file.uri.pathSegments.last}';
      final path = '$userId/$fileName';

      await SupabaseService.client.storage.from('qr_assets').upload(path, file);
      
      // Get public URL
      final publicUrl = SupabaseService.client.storage.from('qr_assets').getPublicUrl(path);
      
      return {'url': publicUrl};
    } catch (e) {
      throw _parseError(e);
    }
  }

  /// Delete a QR code.
  static Future<void> deleteQrCode(String id) async {
    try {
      await SupabaseService.client.from('qr_codes').delete().eq('id', id);
    } catch (e) {
      throw _parseError(e);
    }
  }

  // ── Folders ────────────────────────────────────────────────────────────────

  static Future<List<Folder>> getFolders() async {
    try {
      final userId = SupabaseService.currentUser?.id;
      final data = await SupabaseService.client
          .from('folders')
          .select('*')
          .eq('user_id', userId!);
          
      return data.map((m) => Folder.fromMap(m)).toList();
    } catch (e) {
      throw _parseError(e);
    }
  }

  static Future<Folder> createFolder(String name, {String? iconName}) async {
    try {
      final userId = SupabaseService.currentUser?.id;
      final res = await SupabaseService.client.from('folders').insert({
        'user_id': userId,
        'name': name,
        'icon_name': iconName,
      }).select().single();
      
      return Folder.fromMap(res);
    } catch (e) {
      throw _parseError(e);
    }
  }

  static Future<void> deleteFolder(String id) async {
    try {
      await SupabaseService.client.from('folders').delete().eq('id', id);
    } catch (e) {
      throw _parseError(e);
    }
  }
}

File: web-app/src/lib/services/qrService.ts
import { supabase } from '@/lib/supabase';

export interface QrCodeData {
    id?: string;
    user_id?: string;
    folder_id?: string | null;
    name: string;
    type: string;
    content: Record<string, any>;
    color_primary?: string;
    asset_path?: string | null;
    asset_url?: string | null;
}

/**
 * Service to manage QR Codes synced with Supabase Backend.
 */
export const QrService = {
    /**
     * Upload an asset for a QR code directly to Supabase storage.
     * Enforces the 4.5MB limit on the client side before uploading.
     */
    async uploadAsset(file: File, userId: string): Promise<{ path: string; url: string }> {
        // Enforce 4.5MB Limit
        const MAX_FILE_SIZE = 4.5 * 1024 * 1024;
        if (file.size > MAX_FILE_SIZE) {
            throw new Error('File exceeds the 4.5MB limit. Please compress your file.');
        }

        const fileName = `${userId}/${Date.now()}_${file.name.replace(/[^a-zA-Z0-9.\-_]/g, '')}`;
        
        const { data, error } = await supabase.storage
            .from('qr_assets')
            .upload(fileName, file, {
                cacheControl: '3600',
                upsert: false
            });

        if (error) {
            console.error('Storage Upload Error:', error);
            throw new Error(`Upload failed: ${error.message}`);
        }

        const { data: urlData } = supabase.storage
            .from('qr_assets')
            .getPublicUrl(data.path);

        return {
            path: data.path,
            url: urlData.publicUrl
        };
    },

    /**
     * Save or update a QR code in the database.
     */
    async saveQrCode(qrData: QrCodeData): Promise<string> {
        const payload = {
            folder_id: qrData.folder_id,
            name: qrData.name || `Untitled ${qrData.type} QR`,
            type: qrData.type,
            content: qrData.content,
            color_primary: qrData.color_primary || '#0F172A',
            asset_path: qrData.asset_path,
            asset_url: qrData.asset_url,
            updated_at: new Date().toISOString()
        };

        if (qrData.id) {
            // Update existing
            const { error } = await supabase
                .from('qr_codes')
                .update(payload)
                .eq('id', qrData.id);

            if (error) throw error;
            return qrData.id;
        } else {
            // Insert new
            const { data, error } = await supabase
                .from('qr_codes')
                .insert([{ ...payload, user_id: qrData.user_id }])
                .select('id')
                .single();

            if (error) throw error;
            return data.id;
        }
    },

    /**
     * Fetch all QR codes for a user.
     */
    async fetchQrCodes() {
        const { data, error } = await supabase
            .from('qr_codes')
            .select(`
                *,
                folders (
                    id,
                    name,
                    icon_name
                )
            `)
            .order('created_at', { ascending: false });

        if (error) throw error;
        return data;
    },

    /**
     * Fetch analytics for a specific QR code.
     */
    async fetchAnalytics(qrId: string) {
        const { data, error } = await supabase
            .from('qr_scans')
            .select('*')
            .eq('qr_code_id', qrId)
            .order('scanned_at', { ascending: false });

        if (error) throw error;
        return data;
    }
};
